R/write_output.R

```diff
@@ -0,0 +1,329 @@
+#' combinations for further merging.
+#'
+#' @param fit A Stan fit object with posterior estimates.
+#'
+#' @return A list containing two elements: `stan_draws` (the extracted draws in
+#' long format) and `fact_table` (a table of unique date-time-parameter
+#' combinations).
+#' @noRd
+extract_draws_from_fit <- function(fit) {
+  # Step 1: Extract unique date-time-parameter combinations
+  fact_table <- fit[["estimates"]][["samples"]][,
+    c("date", "time", "parameter"),
+    with = FALSE
+  ]
+  fact_table <- stats::na.omit(unique(fact_table))
+
+  # Step 1.1: Add corresponding 'obs_cases' rows for 'latent_cases' dates
+  obs_fact_table <- fact_table[
+    fact_table[["parameter"]] == "imputed_reports",
+  ]
+  data.table::set(obs_fact_table, j = "parameter", value = factor(
+    obs_fact_table[["parameter"]],
+    levels = c("imputed_reports"),
+    labels = c("obs_reports")
+  ))
+
+  # Combine original fact_table with new 'obs_reports' rows
+  fact_table <- rbind(fact_table, obs_fact_table, fill = TRUE)
+  data.table::setnames(
+    fact_table,
+    old = c("parameter"),
+    new = c(".variable")
+  )
+
+  # Step 2: Extract desired parameters from the Stan object as posterior draws
+  stanfit <- fit[["estimates"]][["fit"]]
+  # Hacky workaround to avoid R CMD check NOTE on "no visible global binding"
+  # for variables in a dataframe evaluated via NSE. To use tidybayes, we need to
+  # use NSE, so giving these a global binding. The standard dplyr hacks
+  # (str2lang, .data prefix) didn't work here because it's not dplyr and we're
+  # not accessing a dataframe.
+  imputed_reports <- obs_reports <- R <- r <- time <- NULL # nolint
+  stan_draws <- tidybayes::gather_draws(
+    stanfit,
+    imputed_reports[time],
+    obs_reports[time],
+    R[time],
+    r[time],
+  ) |>
+    data.table::as.data.table()
+
+  return(list(stan_draws = stan_draws, fact_table = fact_table))
+}
+
+#' Post-process and merge posterior draws with a fact table.
+#'
+#' This function merges posterior draws with a fact table containing
+#' date-time-parameter combinations. It also standardizes parameter names and
+#' renames key columns.
+#'
+#' @param draws A data.table of posterior draws (either raw or summarized).
+#' @param fact_table A data.table of unique date-time-parameter combinations.
+#'
+#' @return A data.table with merged posterior draws and standardized parameter
+#' names.
+#' @noRd
+post_process_and_merge <- function(
+    draws,
+    fact_table,
+    geo_value,
+    model,
+    disease) {
+  # Step 1: Left join the date-time-parameter map onto the Stan draws
+  merged_dt <- merge(
+    draws,
+    fact_table,
+    by = c("time", ".variable"),
+    all.x = TRUE,
+    all.y = FALSE
+  )
+
+  # Step 2: Standardize parameter names
+  data.table::set(merged_dt, j = ".variable", value = factor(
+    merged_dt[[".variable"]],
+    levels = c("imputed_reports", "obs_reports", "R", "r"),
+    labels = c("latent_cases", "obs_cases", "Rt", "growth_rate")
```

Comment:
Yes, I'm making these up.

> latent_final_counts

I don't love this because (1) I think "final" makes sense from the nowcasting perspective but becomes confusing if you're not coming to it thinking about nowcasting. Could we say something more explicit like "nowcast" instead? and (2) I think "count" is unclear because we have incidence, latent cases, and observed cases floating around. It's not clear which one count refers to.

What about "latent_nowcast_cases"?

> latent_obs_counts

I want the @seabbs bat signal to confirm, but I'm pretty sure this isn't correct. The observed counts aren't a latent parameter -- this is the observed parameter being compared to the observed cases in the likelihood evaluation.

The subtlety here is that Sam set things up so that `obs_cases` has _both_ right-truncation and observation (negative binomial) noise. The latent cases have neither. So we want  a name that makes the distinction that these cases have both the right-truncation and the noise applied.

Reply:
I think `obs_reports` is with truncation but without observation noise. `imputed_reports` are posterior predictions for final counts I think. These should have clearer names in the package postprocessing or we can talk through the stan code

`https://github.com/epiforecasts/EpiNow2/blob/2e1aa86598f13364c1af05a5a83b3ac7790db41a/inst/stan/estimate_infections.stan#L242`

Reply:
Talked to Sam this morning and my explanation above is wrong. Took another crack at it here: https://github.com/CDCgov/cfa-epinow2-pipeline/pull/41#discussion_r1787986060

Reply:
I'm going to close this one and keep tracking this in the linked comment.